controller/konnect/reconciler_kongplugin.go

```diff
@@ -242,31 +251,189 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
-type kongPluginBindingMapping struct {
-	byServiceName map[string][]configurationv1alpha1.KongPluginBinding
-	byRouteName   map[string][]configurationv1alpha1.KongPluginBinding
+func deleteKongPluginBindings(
+	ctx context.Context,
+	logger logr.Logger,
+	cl client.Client,
+	pluginBindingsToDelete map[types.NamespacedName]configurationv1alpha1.KongPluginBinding,
+) error {
+	for _, pb := range pluginBindingsToDelete {
+		// NOTE: we check the deletion timestamp here because attempting to delete
+		// and return here would prevent KongPlugin finalizer update.
+		if !pb.DeletionTimestamp.IsZero() {
+			continue
+		}
+		if err := cl.Delete(ctx, &pb); err != nil {
+			if k8serrors.IsNotFound(err) {
+				continue
+			}
+			return err
+		}
+		log.Info(logger, "KongPluginBinding deleted", pb)
+	}
+	return nil
+}
+
+func ownerRefIsPlugin(kongPlugin *configurationv1.KongPlugin) func(ownerRef metav1.OwnerReference) bool {
+	return func(ownerRef metav1.OwnerReference) bool {
+		return ownerRef.Kind == "KongPlugin" &&
+			ownerRef.Name == kongPlugin.Name &&
+			ownerRef.UID == kongPlugin.UID
+	}
 }
 
-func mapKongPluginBindingsByTargetTypeAndRef(bindings []configurationv1alpha1.KongPluginBinding) kongPluginBindingMapping {
-	ret := kongPluginBindingMapping{
-		byServiceName: map[string][]configurationv1alpha1.KongPluginBinding{},
-		byRouteName:   map[string][]configurationv1alpha1.KongPluginBinding{},
+func deleteUnusedKongPluginBindings(
```

Comment:
This function contains quite complex logic that ought to be tested. It's a good candidate for unit tests using a fake cache client I believe.

Reply:
Note: To be clear, I'm fine with adding this in a separate PR.

Reply:
👍 Let's add that in a separate PR together with an envtest based test.